silver/models/transactions.py

```diff
@@ -1,31 +1,219 @@
+
+        if self.invoice and self.proforma:
+            if self.invoice.proforma != self.proforma:
+                raise ValidationError('Invoice and proforma are not related.')
+
     @property
-    def is_usable(self):
-        return not self.disabled and (not self.valid_until or
-                                      self.valid_until > timezone.now())
+    def can_be_consumed(self):
+        if not self.consumable:
+            return False
+
+        if self.valid_until and self.valid_until < timezone.now():
+            return False
+
+        if self.state != Transaction.States.Initial:
+            return False
+
+        return True
 
     @property
     def customer(self):
         return self.payment_method.customer
 
     @property
+    def document(self):
+        return self.invoice or self.proforma
+
+    @property
+    def provider(self):
+        return self.document.provider
+
+    @property
     def payment_processor(self):
         return self.payment_method.payment_processor
+
+    def __unicode__(self):
+        return unicode(self.uuid)
+
+
+def send_email_with_current_state(transaction):
+    if transaction.state == Transaction.States.Initial:
+        send_new_transaction_email(transaction)
+    elif transaction.state == Transaction.States.Pending:
+        send_pending_transaction_email(transaction)
+    elif transaction.state == Transaction.States.Failed:
+        send_failed_transaction_email(transaction)
+    elif transaction.state == Transaction.States.Settled:
+        send_settled_transaction_email(transaction)
+    elif transaction.state == Transaction.States.Refunded:
+        send_refunded_transaction_email(transaction)
+
+
+@receiver(pre_save, sender=Transaction)
+def pre_transaction_save(sender, instance=None, **kwargs):
+    old = get_object_or_None(Transaction, pk=instance.pk)
+    setattr(instance, 'old_value', old)
+
+
+@receiver(post_save, sender=Transaction)
+def post_transaction_save(sender, instance, **kwargs):
+    if not instance.old_value:
+        logger.info('[Models][Transaction]: %s', {
+            'detail': 'A transaction was created.',
+            'transaction_id': instance.id,
+            'customer_id': instance.customer.id,
+            'invoice_id': instance.invoice.id if instance.invoice else None,
+            'proforma_id':
+                instance.proforma.id if instance.proforma else None
+        })
+
+        send_new_transaction_email(instance)
+
+        if instance.state != Transaction.States.Initial:
+            # The transaction was created with a state other than the initial
+            # one
+            send_email_with_current_state(instance)
+
+
+@receiver(post_transition)
+def post_transition_callback(sender, instance, name, source, target, **kwargs):
+    """
+    Syncs the state of the related documents of the transaction with the
+    transaction state
+    """
+    if issubclass(sender, Transaction):
+        if target == Transaction.States.Settled:
```

Comment:
The same here.
Another idea would be to extract this logic in another function, in case that the decision branches will grow.